### ISSUE TITLE
Weird stack trace behaviour

### ISSUE THREAD

**Issue by resnickj** [sebastienros/jint#1202]
I'm having an issue where the JS stack trace appears to accumulate with each error, when previous errors are handled by catching a `JavaScriptException` in CLR code. Here's a test case that reproduces the issue (apologies if it's convoluted, I couldn't find a way to make it any simpler). If you remove the calls to `foo` in the script, the test passes. Each call to `foo` causes the stack trace to accumulate more frames from the previous (handled) errors, causing the test to fail. I'd be willing to submit a PR to fix this, if I could figure out where to start... 

    public class MyTests
    {
        private static void Foo(Func<int> callback)
        {
            try
            {
                var i = callback();
            }
            catch(JavaScriptException)
            {
                // suppress JS errors
            }
        }

        [Fact]
        public void StackTraceBehaviour()
        {
            var engine = new Engine()
                .SetValue("foo", new Action<Func<int>>(Foo));

            const string script = @"
    function throwIt(message) {
        throw new Error(message);
    }

    foo(() => throwIt('e1'));
    foo(() => throwIt('e2'));
    foo(() => throwIt('e3'));
    
    try {
        throwIt('e4');
    } catch(x){
        // return x to host environment
        x;
    }
    ";
            var result = engine.Evaluate(script);
            var stack = result.Get("stack").AsString();
            Assert.Equal("   at throwIt (message) <anonymous>:3:5\r\n   at <anonymous>:11:5", stack);
        }
    }

With calls to `foo` removed, the stack trace is as expected:

    at throwIt (message) <anonymous>:3:5
    at <anonymous>:11:5

With calls to `foo` left in, the observed stack trace ends up looking like this:

    at throwIt (message) <anonymous>:3:5
    at delegate <anonymous>:11:5
    at delegate <anonymous>:8:1
    at delegate <anonymous>:7:1
    at <anonymous>:6:1

**Comment by lahma** [comment]
I think some clues are here:

https://github.com/sebastienros/jint/blob/4b5a32a7aaa58b6d11744af160d07f797aece957/Jint/Runtime/Interpreter/Statements/JintTryStatement.cs#L51-L59

This probably is some kind of combination of need to keep stack trace alive a bit longer, to gather is it when exceptions are thrown. And `ScriptEvaluation` builds it's stack trace too and clears the stack. Interop scenarios might have been overlooked...

**Comment by resnickj** [comment]
Thanks for the tip. Having delved in a bit deeper, I've prototyped a solution based on the following changes:

1. Remove the quick patch that was introduced in #853.
2. Implement the "ideal" fix as described in #853 (i.e. `try/finally` in `Engine.Call` and `Engine.Construct`).
3. Patch the resulting issues that are due to the back and forth conversion between `Completion` -> `JavaScriptException` -> `Completion` -> ...

Although a proper fix to (3) might warrant a deeper re-design (which is beyond my understanding of the Engine), we can resolve the issues superficially at least by simply having the creation of `JavaScriptException` populate the stack trace only if there isn't already a "stack" property present in the `Error` object. The essential code change would be in `JavaScriptException.SetCallstack`, and something like this:

        internal void SetCallstack(Engine engine, Location location, bool overwriteStack)
        {
            Location = location;

            var errObj = Error.IsObject() ? Error.AsObject() : null;
            if (errObj == null)
            {
                _callStack = engine.CallStack.BuildCallStackString(location);
                return;
            }

            // Does the Error object already have a stack property?
            if (errObj.HasProperty(CommonProperties.Stack) && !overwriteStack)
            {
                _callStack = errObj.Get(CommonProperties.Stack).AsString();
            }
            else
            {
                _callStack = engine.CallStack.BuildCallStackString(location);
                errObj.FastAddProperty(CommonProperties.Stack, _callStack, false, false, false);
            }
        }

With this change, and a few other tweaks, all of the existing tests pass, as well as the new test I posted in my initial comment.

This is not a perfect solution by any means, the main limitation (that I'm aware of) being that it relies on the `Error` object as a carrier of the stack-trace between `Completion` -> `JavaScriptException` -> `Completion`. This means if something else is thrown (e.g. `throw "something funny"`), the `JavaScriptException` will report an incorrect stack trace.

Despite this limitation, I think this may be a marginally better (more general) solution than the existing patch committed for #853. Let me know what you think, and if you agree I'll submit a PR.
